### ISSUE TITLE
Inside the C Standard Library

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#4303]
Inside the C Standard Library<br>
<br>
http://bit.ly/2Do58Jr<br>
<br>
About 0 Minutes<br>
<br>
<div>
<div>
<p><img src="https://begriffs.com/images/plauger-c-std.gif" alt="Cover of the Standard C Library"> After diving into the C language through K&amp;R, and then studying portability (see <a href="https://begriffs.com/posts/2018-11-15-c-portability.html">C Portability Lessons from Weird Machines</a>), my next challenge was to take a systematic look at the standard library. To do this I worked through P. J. Plauger&rsquo;s book The Standard C Library (ISBN 978-0131315099) where he examines an implementation of all the functions. It has a chapter for each header, with background information, an excerpt from the C89 standard, tips on use, and full implemention with tests. The author was on the X3J11 committee that defined ANSI C.</p>
<p>As I worked through the book &ndash; trying first to write the examples myself, then comparing his code to mine, and finally running the examples &ndash; I kept notes with questions about portability, rationale, and C behavior. By cross-referencing the following books, asking questions on IRC, and browsing StackOverflow and the comp.lang.c archives, I found satisfactory answers.</p>
<ul>
<li>&ldquo;The C Standard: Incorporating Technical Corrigendum 1&rdquo; by The British Standards Institution (ISBN 978-0470845738). This is the C99 standard itself (rather than C89 like Plauger&rsquo;s book), and it includes an entire first half devoted to the rationale behind language and library choices. This is helpful for understanding C semantics.</li>
<li>&ldquo;Portable C Software&rdquo; by Mark R. Horton (ISBN 978-0138680503). Written after ANSI C was standardized, but early enough where it wasn&rsquo;t fully adopted. He provides early history of each standard library function, as well as some functions that are now defunct.</li>
<li>&ldquo;Portable C&rdquo; by Henry Rabinowitz (ISBN 978-0136859673). Great for illustrating the design decisions of the language as it relates to diverse hardware.</li>
<li>&ldquo;The CERT C Coding Standard&rdquo; by Robert C. Seacord (ISBN 978-0321984043). Illustrates potential insecurity with, among other things, the standard library. Lists real code that caused vulnerabilities.</li>
<li>&ldquo;C Programming FAQs&rdquo; by Steve Summit (ISBN 978-0201845198). I can see why these were historically the most frequently asked questions. I asked many of them myself.</li>
</ul>
<p>This article is not a comprehensive explanation of the standard by any means. It&rsquo;s just things that were new or interesting to me. Some of it may be old news to you, and conversely I may have omitted something that seemed basic to me but would have been useful to mention. The focus is C89, with comparisons to the later standards C99 and C11 when relevant.</p>
<p>(Note: updated several sections based on feedback from Lobste.rs, HN, and Reddit. Thank you for the review, everyone.)</p>
<h3 id="brief-history-of-the-library">Brief History of the Library</h3>
<p>Functions in the library grew organically from communities of programmers sharing ideas and implementations. Many groups of people used C on Unix throughout the 70s, across multiple architectures. They wrote compilers with extra features, and experimented with additions to Unix. By February 1978 core C practice had stabilized to the point where Kernighan and Ritchie codified it in the first edition of their book The C Programming Language (ISBN 978-0131101630).</p>
<p>By 1980 C users formed the &ldquo;/usr/group&rdquo; organization to combine their library experience into an informal standard, which they released in 1984. Meanwhile in 1983, the American National Standards Institute (ANSI) formed a committee, X3J11, to establish a standard specification of C and officially standardize the library. The committee reviewed the work of /usr/group, K&amp;R 1st edition, and various compiler extensions. They deliberated from 1983 to 1989 to produce the C89 standard (&ldquo;ANSI C&rdquo;).</p>
<p>&ldquo;Design by committee&rdquo; may not have pleasant associations for some people, but in this case the committee drew on a lot of experience, and often declined to speculatively innovate, working to clarify existing practice instead. The result was a small, tight language and standard library.</p>
<p>Compared with libraries in other languages, the standard C library is lean. It doesn&rsquo;t have much in the way of general algorithms or containers. This helped the language port easily and more widely. The library has basic facilities for time, math, I/O etc, and operates on simple types. It also provides portable facilities to do non-portable things, like variadic arguments, and non-local gotos.</p>
<p>We will restrict attention to the C standard library only, and not extensions such as POSIX. Thus no functions from unistd.h, fcntl.h etc.</p>
<h3 id="assert.h">assert.h</h3>
<p>A simple way to halt a program with debugging information if an assumption doesn&rsquo;t hold:</p>
<div>
<pre>
<code><span>#include </span><span>&lt;assert.h&gt;</span> <span>int</span> main(<span>void</span>) { assert(<span>1</span> == <span>0</span>); } <span>/* outputs</span> <span>Assertion failed: (1 == 0), function main, file assert.c, line 2.</span> <span>Abort trap: 6</span> <span>*/</span></code>
</pre>
</div>
<p>Because it echoes the statement under test and includes a filename and line number, it&rsquo;s useful for a quick and dirty test suite. As a hack to include a little extra information, logically conjoin a string:</p>
<div>
<pre>
<code>assert(<span>1</span> == <span>0</span> &amp;&amp; <span>"Checking this because..."</span>);</code>
</pre>
</div>
<p>Furthermore, assert() calls abort() rather than exit(), which causes the program to quit and dump core if permitted by the operating system. If the binary was compiled with debugging support you can load that core file in a debugger and inspect the program&rsquo;s full state at the time the assertion was made:</p>
<div>
<pre>
<code><span>gdb</span> -c /path/to/corefile <span># ^^^ inspect variables, backtrace, etc in the debugger</span></code>
</pre>
</div>
<p>This means that littering functions liberally with assertions can be a good way to debug problems. It provides richer information than debug print statements in situations where it&rsquo;s OK to terminate the program. It also adds no overhead to the final release, because the assertions will be removed by the preprocessor when compiled with NDEBUG defined. This can be done by adding <code>-DNDEBUG</code> to CFLAGS or by adding a regular <code>#define NDEBUG</code> in the code.</p>
<p>All headers in the standard library are idempotent, except assert.h. By including it multiple times in a file you can enable or disable the assert macro.</p>
<div>
<pre>
<code><span>#define NDEBUG</span> <span>#include </span><span>&lt;assert.h&gt;</span> <span>/* Now assert() won't do anything... */</span> <span>#undef NDEBUG</span> <span>#include </span><span>&lt;assert.h&gt;</span> <span>/* Now assert() works again */</span></code>
</pre>
</div>
<p>It&rsquo;s OK to include assert.h twice in a row because it causes what C calls a &ldquo;benign&rdquo; redefinition of the assert macro.</p>
<div>
<pre>
<code><span>/* no harm, this is a benign redefinition: */</span> <span>#define FUN 1</span> <span>#define FUN 1</span> <span>/* not benign and not allowed: */</span> <span>#define FUN 1</span> <span>#define FUN 2</span></code>
</pre>
</div>
<p>A properly designed assert macro should work in any context, even somewhere weird like <code>for (i = 0, assert(n&lt;10); i &lt; n; ++i)</code>. Because I didn&rsquo;t think of this, my own attempt at writing assert used an if statement. In fact this is what Mark Horton shows in his Portable C Software book:</p>
<div>
<pre>
<code><span>/* incorrect definition */</span> <span>#ifndef NDEBUG</span> <span>#define assert(p) (if(!(p)) ... )</span> <span>#else</span> <span>#define assert(p) ;</span> <span>#endif</span></code>
</pre>
</div>
<p>This can be improved. Plauger uses the ternary operator, and substitutes the result <code>((void)0)</code> when the predicate holds. He also keeps the code (to print the error and exit) in a real function which is defined in a separate source file. That&rsquo;s because headers in the standard library are not allowed to include each other. It&rsquo;s a self-imposed discipline.</p>
<p>The last trick I found interesting was how to delay preprocessor evaluation into two steps using these helpers:</p>
<div>
<pre>
<code><span>/* a "thunk" to evaluate __LINE__ */</span> <span>#define _STR(x) _VAL(x)</span> <span>#define _VAL(x) #x</span></code>
</pre>
</div>
<p>The string can then be built as</p>
<div>
<pre>
<code><span>"Assertion failed: "</span> #p <span>", file "</span> __FILE__ <span>", line "</span> _STR(__LINE__) <span>"</span><span>\n</span><span>"</span></code>
</pre>
</div>
<p>Finally, some implementations tolerate an arbitrary scalar expression as the argument to assert, but the ANSI committee decided to require int expressions for correct operation.</p>
<p>Given that it was the first header I saw in Plauger&rsquo;s book, this is where I learned that the standard reserves names like _Foo starting with underscore and a capital letter, for the standard library. Don&rsquo;t use that naming convention in regular code.</p>
<h3 id="ctype.h">ctype.h</h3>
<p>Although I learned a few things from this header, it was ultimately kind of a letdown. It&rsquo;s not sufficient for international text processing because, while the functions operate on the int type, they specify that its value should fall in the range of unsigned char or the negative value EOF. So while non-English speakers can use a new codepage for their locale, it will be unable to hold more than 255 symbols (plus <code>\0</code>, plus EOF), which is too few for east asian languages. C99 introduced wctype.h to operate on wide characters, but that has its own problems. (More about that in the stdlib.h section below.)</p>
<p>Even for languages which fit in an 8-bit codepage, the ctype functions don&rsquo;t always suffice. For example the German letter &szlig; uppercases into two letters, SS. The tupper() function in ctype.h can&rsquo;t handle it, replacing one character with another. The greek letter &Sigma; ordinarily lowercases to &sigma;, except at the end of a word where it should be &sigmaf;. The tolower() function doesn&rsquo;t have enough context to pick the correct form.</p>
<p>I still learned some interesting techniques by studying this header. Plauger implements all the isxxxxx() functions with a lookup table of bit-packed shorts. The table itself is specific to EOF (from stdio.h) being -1, and relies on a certain size of unsigned char. The code uses a nice trick to fail early on a system where this assumption is incorrect:</p>
<div>
<pre>
<code><span>#include </span><span>&lt;limits.h&gt;</span> <span>#include </span><span>&lt;stdio.h&gt;</span> <span>#if EOF != -1 || UCHAR_MAX != 255</span> <span>#error WRONG TABLES IN CTYPE.H</span> <span>#endif</span></code>
</pre>
</div>
<p>Although the code is non-portable, it fails in the most honest and upfront way possible at compile time. Also as a historical note, the isxxxxx() functions used to be defined only for 7-bit characters, but ANSI requires them to handle all values for unsigned char.</p>
<p>One advantage of using a lookup table is that the value to be looked up requires only one evaluation. Thus the lookup function can be a macro without danger of executing code with a side effect more than once.</p>
<div>
<pre>
<code><span>/* evaluates c at least twice</span> <span> * (also ASCII-specific, but don't focus on that)</span> <span> */</span> <span>#define isalpha(c) \</span> <span> (((c) &gt;= 'a' &amp;&amp; (c) &lt;= 'z') || ((c) &gt;= 'A' &amp;&amp; (c) &lt;= 'Z'))</span> <span>/* evaluates c only once */</span> <span>#define isalpha(c) (_Ctype[(int)(c)] &amp; (_LO|_UP))</span> <span>/* consider how this would behave */</span> <span>if</span> (isalpha(c = getchar())) ...</code>
</pre>
</div>
<p>Using macros for these little functions is good for performance. However, every library function in the standard library (unless specifically noted otherwise) must be represented as an actual function too, in case a program wishes to pass its address as a parameter to another function. How can we define a function called &ldquo;isalpha&rdquo; if that name is also recognized by the preprocessor? I learned you can just enclose the name in parens:</p>
<div>
<pre>
<code><span>int</span> (isalpha)(<span>int</span> c) { <span>return</span> (_Ctype[(<span>int</span>)(c)] &amp; (_LO|_UP)); } <span>/* now &amp;isalpha will be defined if needed */</span></code>
</pre>
</div>
<p>The trick is used throughout the standard library but I first saw it in ctype. Conversely, every library function is a candidate for redefinition as a macro, provided that the macro evaluates each of the arguments exactly once and parenthesizes them thoroughly. (Well, getc is an exception but more on that later.)</p>
<p>Another trick is to perform an array lookup using a pointer that is shifted forward in the original array.</p>
<div>
<pre>
<code><span>static</span> <span>const</span> <span>short</span> ctyp_tab[<span>257</span>] = {<span>0</span>, <span>/* ... */</span>}; <span>const</span> <span>short</span> *_Ctype = &amp;ctyp_tab[<span>1</span>];</code>
</pre>
</div>
<p>The shifted pointer allows EOF (-1) to be looked up easily without undefined behavior. The expression <code>_Ctype[EOF]</code> means <code>_Ctype[-1]</code> which is the same as <code>*(ctyp_tab+1-1)</code>, which does not attempt to dereference &ndash; or even point to &ndash; memory before a primary data object. Pointers are allowed to be assigned only addresses either inside, or one space to the right, of a data object. (Data objects are arrays, structures, or regions allocated from the heap. See Rabinowitz&rsquo;s book for a good discussion of this.)</p>
<p>Character codes 128-255 are interpreted as negative numbers when considered as signed char. C does not specify whether char is signed or unsigned (it varies by platform). When char is signed, beware of converting it to int for the ctype functions. The integral promotion will &ldquo;sign-extend,&rdquo; creating a negative int value. This is not suitable for ctype functions, which require int values that are either storable in unsigned char or are the special EOF value. To avoid sign extension, cast char values to unsigned char in calls to ctype functions.</p>
<h3 id="errno.h">errno.h</h3>
<p>Errno is the mechanism everyone loves to hate. The X3J11 committee wanted to remove it but decided not to make such a radical innovation on existing practice. In the early drafts of the standard it was kept it in stddef.h, but they decided that stddef should exist on freestanding environments and they split errno.h off into its own header.</p>
<p>Not much to say about it. The errno global variable is set to zero at the start of program execution. Library functions can set it to nonzero values but will never set it to zero themselves. To use it, explicitly set it to zero yourself, call a library function, then check whether errno changed.</p>
<p>One interesting tidbit is that errno is sometimes not a global variable at all, but a macro for <code>(*_Error())</code>. This allows different program threads to see a different value in a multi-threaded application. Even before threading considerations, having to set a real data object immediately after performing hardware floating point ops would break the FPU pipeline. Allowing the check to be deferred until requested with this _Error() function doesn&rsquo;t break the pipeline.</p>
<h3 id="float.h-limits.h-and-math.h">float.h, limits.h, and math.h</h3>
<p>Both float.h and limits.h are inventions of the committee. You can generate them with the <a href="https://homepages.cwi.nl/~steven/enquire.html">enquire</a> program written by Steven Pemberton. It performs runtime checks on the data types to find information about them, then generates the desired header file. It&rsquo;s very (but not completely) portable, as it &ldquo;only works if overflows are ignored by the C system or are catchable with signal().&rdquo; It also detects and outputs information about the representation of base types, like endianness.</p>
<p>I decided not to slow down to study these headers in depth because I lack the knowledge of floating point representation necessary to understand the internals of the functions inside. I do know the definitions and parameters in float.h were recommended by numerical analysts on the Committee. The set was chosen so as not to prejudice an implementation&rsquo;s selection of floating-point representation.</p>
<p>The math functions are written carefully to avoid overflow and underflow. I&rsquo;ll revisit the topic after studying Michael Overton&rsquo;s short book, &ldquo;Numerical Computing with IEEE Floating Point Arithmetic.&rdquo;</p>
<h3 id="locale.h">locale.h</h3>
<p>Changing a program&rsquo;s locale tells it how to handle local customs. There are multiple locale categories that can be adjusted independently. They control different things, like the codeset used by ctype, the date or monetary formatting desired, or alphabetical sort order. Typically all categories are set to the same locale. The alternative &ndash; a so called &ldquo;mixed locale&rdquo; &ndash; is less common.</p>
<p>Some of the settings are meant for interpretation by your own program, and others automatically affect the C standard library. For example, when parsing a double, strtod checks what the current locale uses for a decimal point symbol. Even if you want to avoid ctype and use a third-party Unicode library, some of the locale information is still useful for your program.</p>
<p>By default C programs use the &ldquo;C&rdquo; locale, which is ASCII text and American formatting. The most respectful thing, though, is to accept the locale in all categories as set by system environment variables. This is indicated by the empty string for locale name.</p>
<div>
<pre>
<code><span>#include </span><span>&lt;locale.h&gt;</span> <span>#include </span><span>&lt;stdio.h&gt;</span> <span>#include </span><span>&lt;stdlib.h&gt;</span> <span>int</span> main(<span>void</span>) { <span>if</span> (setlocale(LC_ALL, <span>""</span>) == NULL) { fputs(<span>"Unable to select system locale"</span>, stderr); <span>return</span> EXIT_FAILURE; } <span>/* ... */</span> }</code>
</pre>
</div>
<p>To see what locales are known on your local system, run</p>
<p>There are 203 installed on MacOS. The list begins with:</p>
<ul>
<li>en_NZ</li>
<li>nl_NL.UTF-8</li>
<li>pt_BR.UTF-8</li>
<li>fr_CH.ISO8859-15</li>
<li>eu_ES.ISO8859-15</li>
<li>en_US.US-ASCII</li>
<li>af_ZA</li>
<li>&hellip;</li>
</ul>
<p>They are in the format [language[_territory][.codeset]]. OpenBSD has chosen to support only the &ldquo;C&rdquo; (aka &ldquo;POSIX&rdquo;) and &ldquo;UTF-8&rdquo; codesets, but supports many languages and territories in the other locale categories. Unicode subsumes all those partial character encodings, so BSD just wanted to eliminate a source of complexity.</p>
<p>One way to see locales in action is by setting environment variables and using Unix tools. We can change sort order using LC_COLLATE.</p>
<div>
<pre>
<code><span>cat</span> <span>&lt;&lt;EOF</span> <span>&gt;</span><span>cn.txt</span> &#25105; &#21916; &#27489; &#25972; &#29702; &#21934; &#35422; EOF</code>
</pre>
</div>
<table>
<thead>
<tr>
<th>LC_COLLATE=C sort &lt;cn.txt</th>
<th>LC_COLLATE=zh_CN.UTF-8 sort &lt;cn.txt</th>
</tr>
</thead>
<tbody>
<tr>
<td>&#29702;</td>
<td>&#21916;</td>
</tr>
<tr>
<td>&#25105;</td>
<td>&#21934;</td>
</tr>
<tr>
<td>&#35422;</td>
<td>&#25105;</td>
</tr>
<tr>
<td>&#27489;</td>
<td>&#25972;</td>
</tr>
<tr>
<td>&#25972;</td>
<td>&#27489;</td>
</tr>
<tr>
<td>&#21916;</td>
<td>&#29702;</td>
</tr>
<tr>
<td>&#21934;</td>
<td>&#35422;</td>
</tr>
</tbody>
</table>
<p>Setlocale() doesn&rsquo;t play well with threads because it updates an internal global variable. Set the locale before spawning threads.</p>
<p>One other random nugget of wisdom from the book, unrelated to locales but mentioned in that chapter, is to use a tool to visualize the call tree in a C program. This can help you understand a new codebase. Try <a href="http://pubs.opengroup.org/onlinepubs/9699919799/utilities/cflow.html">cflow</a>. See also Steve Summit&rsquo;s C FAQ question <a href="http://c-faq.com/resources/tools.html">18.1</a>.</p>
<h3 id="setjmp.h">setjmp.h</h3>
<p>The setjmp/longjump functions create a &ldquo;saveable goto&rdquo; statement to return to places you&rsquo;ve already been. It allows jumping from one function into another. They are like the C programmer&rsquo;s version of exception handling. Setjmp and longjmp are very tricky inside, as they have to save and restore variables, arguments, registers etc to resume execution in another location.</p>
<div>
<pre>
<code><span>#include </span><span>&lt;setjmp.h&gt;</span> <span>#include </span><span>&lt;stdio.h&gt;</span> <span>void</span> dostuff(<span>void</span>); jmp_buf target; <span>int</span> main(<span>void</span>) { <span>if</span> (setjmp(target) == <span>0</span>) { puts(<span>"Saved the target, continuing on."</span>); dostuff(); } <span>else</span> puts(<span>"I feel like I've been here before..."</span>); <span>return</span> <span>0</span>; } <span>void</span> dostuff(<span>void</span>) { longjmp(target, <span>42</span>); }</code>
</pre>
</div>
<p>To indicate that the jmp_buf was set, setjmp returns 0. When execution is jumped back to this point, setjmp returns value passed as the second argument to longjmp, for us 42. The same if statement is evaluated again but with a different result, like waking up in an alternate universe.</p>
<p>As simple as this looks, it&rsquo;s easy to break. The statement containing setjmp must be very simple. Calling setjmp in an if statement or switch statement are fine, but you should not save the return value like <code>n = setjmp(...)</code>. An assignment statement is too complicated and can disturb the sensitive machinery.</p>
<p>The function containing setjmp should be as simple as possible. Only variables declared as volatile are guaranteed to be restored. Generally it&rsquo;s best to execute the real processing in another function called when setjmp returns 0.</p>
<p>Longjmp should not be called from an exit handler (i.e., a function registered with the atexit function). Finally it&rsquo;s undefined behavior to attempt to longjmp to a function that has since returned. Thus longjmp&rsquo;s usual use case is like exceptions in other languages, going back up the call chain, skipping intermediate functions.</p>
<p>The jmp_buf type is actually an array behind a typedef, which is why it&rsquo;s typically passed to setjmp without an ampersand. The standard forbids jmp_buf to be implemented as a scalar or struct.</p>
<p>Given all these caveats, the committee considered requiring that compilers recognize calling setjmp as a special case. Then the function could work in all types of statements. However they decided against it for consistency because they don&rsquo;t require any other function to be a special case (although they allow compiler writers to make special cases as desired).</p>
<h3 id="signal.h">signal.h</h3>
<p>Signals are a UNIX technique for interprocess communication that causes a process to asynchronously call a handler function, or else take a default action. Programs also receive &ldquo;synchronous&rdquo; signals for their own logical exceptions like division by zero, segmentation faults, or floating point problems.</p>
<p>The ANSI committee decided to standardize a weakened portable version of signal functionality. Portable signal handlers can do very little. Here&rsquo;s a typical example of what a handler can do safely:</p>
<div>
<pre>
<code><span>/* a global */</span> <span>volatile</span> <span>sig_atomic_t</span> intflag = <span>0</span>; <span>/* SIGINT handler */</span> <span>void</span> field_int(<span>int</span> sig) { signal(SIGINT, &amp;field_int); intflag = <span>1</span>; <span>return</span>; }</code>
</pre>
</div>
<p>It does as little as possible, simply setting a global variable which regular code can check at leisure. One thing to note is that it re-installs itself in the very first line with the signal() function. That&rsquo;s because on some platforms (like Linux), as soon as a signal is handled it reverts to its default handler, which in the case of SIGINT terminates the program. Other systems such as BSD leave a handler installed when called.</p>
<p>On Linux this would be an unwise thing to do:</p>
<div>
<pre>
<code><span>void</span> field_int_badly(<span>int</span> sig) { <span>/* open a window where a repeat signal could</span> <span> * hit the default handler before we reinstall */</span> sleep(<span>1</span>); signal(SIGINT, &amp;int_catch); intflag = <span>1</span>; <span>return</span>; }</code>
</pre>
</div>
<p>Even our earlier technique of calling signal() right away in the handler isn&rsquo;t completely safe. The CERT C Coding Standard warns that this leaves a tiny window open for a race condition. They suggest not to use the C library signal functionality at all, but the equivalent POSIX functions instead. POSIX allows you to specify persistence of the handler during initial registration.</p>
<p>Another thing to note is the type declaration of the shared flag (that we called intflag in the example). Exception handlers should read and write only volatile variables. For asynchronous exceptions, volatile alone isn&rsquo;t even enough. The variable should be small enough that it can be read or written atomically by the processor.</p>
<p>The C standard provides the sig_atomic_t typedef for this. Each standard library implementation defines it as an alias for a suitably small integral underlying type. If you&rsquo;re writing portable code, don&rsquo;t assume that sig_atomic_t is anything bigger than a char, and don&rsquo;t assume its signedness. Thus the portable value range is 0&hellip;127, although C99 added macros to determine its min and max values.</p>
<p>The standard says not to call any standard library functions from a signal handler except abort(), exit(), longjmp() or signal(). Certainly avoid any functions that interact with state, like those performing I/O or else stdio streams can become corrupted.</p>
<p>Even though the C standard says it&rsquo;s OK to call longjmp from a signal handler, CERT gave an example where doing so caused a vulnerability [VU #834865] in Sendmail because it allowed attackers to time a race condition in main() by timing signals.</p>
<p>A program can raise signals for itself with the raise() function. (It used to be called kill.) However longjmp() is less tricky than raising signals for yourself and should be preferred.</p>
<p>The standard library defines this limited list of signals:</p>
<ul>
<li>SIGABRT abnormal termination, such as is initiated by the abort function</li>
<li>SIGPPE an erroneous arithmetic operation, such as zero divide or an operation resulting in overflow</li>
<li>SIGILL detection of an invalid function image, such as an illegal instruction</li>
<li>SIGINT receipt of an interactive attention signal</li>
<li>SIGSEGV an invalid access to storage</li>
<li>SIGTERM a termination request sent to the program</li>
</ul>
<p>Using other signals makes a program less portable.</p>
<h3 id="stdarg.h">stdarg.h</h3>
<p>On the PDP-11 it was easy to walk function arguments with a pointer. The memory layout of arguments was well known, the size of pointers was the same as the size of int, and the original C language could not pass structures by value.</p>
<p>However when spreading to other architectures, C benefited from creating a portable way to access variable numbers of arguments. Non-PDP architectures had complex calling conventions. Using a library for variadic arguments makes code clearer too, whether or not portability is an issue.</p>
<p>Pre-ANSI C on UNIX used &lt;varargs.h&gt;, which required a final &ldquo;dummy argument.&rdquo; ANSI C got more picky about arguments matching a declaration, and introduced the &ldquo;&hellip;&rdquo; token to take the place of the dummy argument, which breaks varargs. The &ldquo;&hellip;&rdquo; also signals to a compiler that it may want to change the function calling convention.</p>
<p>The committee turned varargs.h into stdarg.h, and generalized the macros to the extent that all known C implementations would be able to handle them with little modification. This functionality was important for other functions in the standard library like printf and scanf.</p>
<p>Using the library is pretty easy, just initialize the list based on the final fixed argument, loop through the args, and release the list.</p>
<div>
<pre>
<code><span>/* add a list of n numbers */</span> <span>int</span> sum(<span>int</span> n, ...) { va_list ap; <span>int</span> s; va_start(ap, n); <span>for</span> (s = <span>0</span>; n &gt; <span>0</span>; --n) s += va_arg(ap, <span>int</span>); va_end(ap); <span>return</span> s; }</code>
</pre>
</div>
<p>It is undefined behavior to call va_arg() more times than there are arguments, so the function will need to determine the number of arguments by other means. Our function above consults the n parameter for that information.</p>
<p>Speaking of the <em>n</em> variable, we&rsquo;re not actually passing its <em>value</em> to va_start(), much as it may look. The va_start macro manipulates &ldquo;n&rdquo; as merely a name, so it can calculate the address of the next argument. Both va_start and va_arg must be implemented as macros, not as functions.</p>
<p>These are the portable assumptions for using stdarg.h:</p>
<ul>
<li>The variadic function must declare at least one fixed argument</li>
<li>The function must call va_end before returning (for cleanup on some architectures)</li>
<li>va_arg can deal only with those types that become pointers by appending &ldquo;*&rdquo; to them. Thus register variables, functions, and arrays can&rsquo;t be returned by va_arg</li>
<li>If a type widens with default argument promotions, then va_arg should request the widened type</li>
</ul>
<p>The last point requires some explanation. When functions had no prototypes in pre-ANSI C, the compiler would promote smaller types to wider ones when sending them to functions. That&rsquo;s because doing so is essentially free &ndash; it costs more to put a byte in a register than to put a word in.</p>
<p>Although ANSI requires functions to have prototypes, the promotion rule still applies to variadic arguments. Char and short get expanded to int, and float gets promoted to double. Thus to accept a char argument, ask for its value with <code>va_arg(ap, int)</code> and then cast to char. Don&rsquo;t do <code>va_arg(ap, char)</code>.</p>
<p>To pass a va_list to another function for continued processing either a) memcpy it if you want to consume all arguments in the current function or b) pass a pointer to it if the called function should consume some or all of the arguments. C99 added va_copy() for the first scenario.</p>
<p>The implementation of stdarg.h macros is gnarly and entirely platform specific. On my system they resolve to builtin compiler functions.</p>
<h3 id="stddef.h">stddef.h</h3>
<p>Stddef.h is a catchall place for definitions, sort of like stdlib.h. Why make two headers rather than combine them into one? It&rsquo;s because C can be compiled in either a &ldquo;hosted&rdquo; or a &ldquo;freestanding&rdquo; environment. The latter is for embedded programming where there isn&rsquo;t enough room for the entire standard library. An implementation must include all the standard library headers to be considered a hosted environment, while a freestanding environment must include only float.h, limits.h, stdarg.h, and stddef.h.</p>
<p>The committee deliberated about putting the things in stddef into the C language itself, but decided the need to extend C is not quite there.</p>
<p>Stddef provides ptrdiff_t, size_t, wchar_t, NULL, and offsetof(structname, attrib). Only ptrdiff_t and offsetof are unique to this header. Other headers usually contain duplicates of the other definitions.</p>
<p>I used to think that NULL was a reserved word in C, but have come to learn that the constant 0 is actually the crux. The compiler treats 0 specially in a pointer context and transforms it to whatever value represents the NULL pointer on the given architecture (which needn&rsquo;t be bitwise zero). Thus NULL is typically a macro for <code>((void *)0)</code>. Ordinarily function pointers and data pointers cannot be mixed, but the standard makes an exception for the null pointer constant, allowing it to be assigned to even a function pointer: <code>int (*foo)(void) = NULL</code>.</p>
<p>The typedef size_t is an unsigned integral type big enough to hold the size of the largest possible object in memory. In some systems that might not be very large, for instance only 64k in the segmented memory model on the Intel 80286. Related rule of thumb: if a variable is going to index an array, it should be type size_t.</p>
<p>The typedef ptrdiff_t is a signed integral type of the result of a pointer subtraction. It&rsquo;s signed because if (p - q) is positive then (q - p) will be negative. Note that if size_t is already the largest integer type, then ptrdiff_t can be no larger, yet the latter loses one bit to hold the sign. So it&rsquo;s possible to make an array with cells too far apart for ptrdiff_t to measure. (Assuming there is room in memory for such a large single array.)</p>
<p>C99 provides a macro SIZE_MAX with the maximum value possible in size_t. C89 doesn&rsquo;t have it, although you can obtain the value by casting <code>(size_t)-1</code>. This works in any number representation (C89 section 6.2.1.2, C99 section 6.3.1.3 signed and unsigned integers), because the value (-1) is converted by &ldquo;repeatedly adding or subtracting one more than the maximum value that can be represented in the new type until the value is in range of the new type.&rdquo;</p>
<p>The offsetof() macro can determine the offset in bytes of a member within its structure. This cannot easily be determined otherwise due to structure padding. The C99 rationale talks about using it to provide &ldquo;generic&rdquo; routines that work from descriptions of the structures, rather than from the structure declarations themselves. On many platforms it is defined as:</p>
<div>
<pre>
<code><span>#define offsetof(type, field) ((size_t)(&amp;((type *)0)-&gt;field))</span></code>
</pre>
</div>
<p>That&rsquo;s undefined behavior, but that&rsquo;s what the standard library is for: a portable way to do sometimes non-portable operations.</p>
<h3 id="stdio.h">stdio.h</h3>
<p>UNIX I/O was clean and simple compared with other systems of its day. It was 8-bit clean, and used a consistent line terminator. At the edges ioctl() would translate the simple I/O streams for the idiosyncrasies of attached devices. The kernel would hold file control state internally and give programs a simple integer descriptor for to use when reading and writing.</p>
<p>Leaving UNIX, C ran up against the complexity of I/O on other systems. The X3J11 committee talked with vendors and came away with a sharper understanding of the I/O model they wanted to support. They had to distinguish text and binary modes. While UNIX used &lsquo;&rsquo; for line endings, DOS used <code>\r\n</code> and Mac classic used &lsquo;&rsquo;. These endings have to be normalized to &lsquo;&rsquo; in text mode but not binary mode. UNIX ignores binary mode, but you better enable it for portability when necessary.</p>
<p>UNIX also had an unusually faithful representation of files. You could put bytes in and expect to read them out unchanged. When doing fully portable I/O keep these caveats in mind:</p>
<ul>
<li>A final line without a terminating newline (in text mode) can confuse some systems, and they may drop the line or append a newline.</li>
<li>Don&rsquo;t count on the system preserving trailing space in a line. Some systems strip it out. Conversely, some systems <em>add</em> a space to a blank line so the line &ldquo;has something in it.&rdquo;</li>
<li>The maximum fully portable line length is 254 characters.</li>
<li>Implementations are free to pad the end of binary files with a block of NUL characters to make the files match certain disk block sizes.</li>
</ul>
<p>Another difference between the C standard I/O and UNIX is buffering. In UNIX, people often wrote their own buffering code to reduce the number of relatively costly I/O system calls. The X3J11 committee decided to include this buffering functionality in stdio.</p>
<p>Buffering is an optimization that can be tailored to expected patterns of I/O. The standard library provides the setvbuf() function to change the size and location of a stream&rsquo;s buffer, as well as choosing between line or block buffering. By default, stdin and stdout are line and block buffered respectively, and stderr is unbuffered. Setvbuf() must be called immediately after a stream is opened, before I/O happens, to have any chance of working.</p>
<h4 id="stdio.h-opening-and-closing">(stdio.h) opening and closing</h4>
<p>Perhaps surprisingly, there is a lot to learn about just opening files. First there may be a limit on how long a filename can be on a system. Stdio provides the FILENAME_MAX macro with this limit. If the system imposes no practical limit then the macro is just a suggested size. This value could be both too short, or paradoxically too long. If it is set very large then you might end up wasting memory or causing problems if allocating on the stack.</p>
<p>Similarly L_tmpnam is the size needed for an array large enough to hold a temporary filename generated by tmpnam(). This function is a security hazard (though it can be useful for generating entropy). It introduces a Time of Check, Time of Use (TOCTOU) race condition because another program or thread could obtain the same temporary file name and create the file first. Use the tmpfile() function instead which actually creates the file, and registers it for removal on normal program exit().</p>
<p>Another common TOCTOU happens with fopen() when trying to create but not replace a file. Programs first check existence, right after which an evildoer can create a symlink of the same name in time for the fopen with &ldquo;w&rdquo; mode to overwrite another file with possibly elevated permissions.</p>
<div>
<pre>
<code><span>/* dangerous */</span> FILE *fp = fopen(<span>"foo.txt"</span>,<span>"r"</span>); <span>/* &lt;-- attacker gets busy here */</span> <span>if</span> (!fp) { fp = fopen(<span>"foo.txt"</span>,<span>"w"</span>); ... fclose(fp); } <span>else</span> fclose(fp);</code>
</pre>
</div>
<p>C11 fixes this with an &ldquo;x&rdquo; (exclusive) mode modifier. If exclusive mode is specified (&ldquo;wx&rdquo;), the fopen fails if the file already exists or cannot be created. In C89 you can either go beyond the standard library, using the POSIX <code>open()</code> function with the <code>O_CREAT | O_EXCL</code> flags, or just try to keep the time between check and write as small as possible.</p>
<p>Once you have opened a file to your liking, or have been given a FILE pointer, treat the pointer as totally opaque. Don&rsquo;t even try to make a copy of the FILE structure because some implementations rely on magic memory addresses for some of them. The CERT standard (FIO38-C) says that the following can cause a crash on some platforms if you try to use <code>my_stdout</code>:</p>
<div>
<pre>
<code><span>/* don't do this */</span> FILE my_stdout = *stdout;</code>
</pre>
</div>
<p>There&rsquo;s a related function called <code>freopen()</code>, but it&rsquo;s not used very often. The main use is converting a big program from reading stdin to reading a named file. It&rsquo;s the simplest way to do that, whereas a new program should just directly fopen whatever file it wants.</p>
<p>During a normal program exit, all open files will be closed. Still, it&rsquo;s useful to explicitly call fclose() on file handles. It helps avoid exceeding the FOPEN_MAX limit of files that can be open at once. Also, failing to properly close files may allow an attacker to exhaust system resources and can increase the risk that file buffers will not be flushed in the event of abnormal program termination.</p>
<p>Speaking of flushing buffers, fflush() <em>might</em> force items in the buffer to be processed, but there is no guarantee. For convenience, fflush(NULL) flushes all streams, which is useful in preparation for possible loss of program control, like going into a dangerous section, or telling the user to turn off the computer.</p>
<p>Two other quirks. Some operating systems will not actually create a file that you fopen() and fclose() unless you write something. Also you can close stdout or stderr, and there are sometimes <a href="https://www.gnu.org/ghm/2011/paris/slides/jim-meyering-goodbye-world.pdf">reasons</a> to do so!</p>
<h4 id="stdio.h-file-navigation">(stdio.h) file navigation</h4>
<p>Stdio.h has two similar pairs of functions to move around in a file: fgetpos/fsetpos and ftell/fseek. Why the duplication? The second pair represents position as a long integer. When the file is opened in binary mode, this long is the number of bytes from the start of file. This is useful because you can do arithmetic on the integer to jump to particular places. The drawback is that on some systems a long is only 32 bits, so cannot support large files.</p>
<p>The fgetpos/fsetpos pair works using a special structure that can represent positions in huge files. You must treat treat this structure as a magic bookmark. It&rsquo;s only obtainable from a call to fgetpos, you cannot construct your own to point to a position you haven&rsquo;t already been.</p>
<p>Stdio also includes a rewind() function, but don&rsquo;t use it. It actively clears the error indicator for a stream. Instead do a <code>fseek(stream, 0L, SEEK_SET)</code>.</p>
<p>These navigation commands were interesting to me, so I created a project called <a href="https://github.com/begriffs/randln">randln</a> to experiment with different ways of picking a random line from a text file. You might find it interesting to look at the pros and cons of each method as explained in the readme.</p>
<p>You can actually inspect I/O as it happens. The trick is walk the program through a debugger while tracing its system calls in another terminal. To use that randln program as an example, First start it in the debugger. Enable tracing for the debugger&rsquo;s I/O and that of its children.</p>
<p>Put a breakpoint in the line-finding function and start the program. In another terminal find the PID of randln that was launched by gdb, and start tailing the ktrace dump. This command will show the first 20 bytes of data in each I/O request:</p>
<p>Now step through the program in the debugger and watch the consequences of each statement.</p>
<p>A final note about stream navigation and the shell. I noticed when doing <code>foo &lt;bar</code> the program can perform fseeks on bar, but <code>cat bar | foo</code> cannot. It&rsquo;s another reason not to abuse cat.</p>
<h4 id="stdio.h-reading-and-writing">(stdio.h) reading and writing</h4>
<p>The foundation of all stream input is fgetc(). The other standard library input functions must be implemented <em>as if</em> they call it repeatedly, even if they don&rsquo;t. It pulls a character out of a stream (or a character that was pushed back by ungetc() if such exists), and refills the stream buffer if needed. Some platforms allow ungetc() to push back a whole stack of characters, but the portable assumption is that it can store only one.</p>
<p>While fgetc is a function, getc() is a macro that avoids incurring a function call just to get a character. The downside is that getc() is allowed to evaluate its argument more than once, so don&rsquo;t do anything with a side effect there. Now, fgetc() is allowed to be a macro too, but may not be as efficient because it has less freedom in its implementation. It is not permitted to evaluate its argument twice.</p>
<p>Moving up the food chain we come to gets(), which reads characters into a string until it encounters the NUL character &ndash; and it&rsquo;s a buffer overflow waiting to happen. It was removed in C11. The fgets() function is better because you can specify a max length. However if it fails, the contents of the array being written is indeterminate. It is necessary to reset the string to a known value to avoid errors on subsequent string manipulation functions.</p>
<p>The fread/fwrite functions work in chunks. My only notes about them are</p>
<div>
<pre>
<code><span>/* prefer this */</span> fread(buf, <span>1</span>, size*n, stream) <span>/* over this */</span> fread(buf, size, n, stream)</code>
</pre>
</div>
<p>The second form is worse because you can&rsquo;t detect whether it read an extra size-1 characters past what it reports. Also some implementations of fread (fwrite) simply call fgetc (fputc) in a loop, whereas others are more optimized. Doing a straight UNIX read (write) can be faster.</p>
<p>The standard library doesn&rsquo;t allow alternating reads and writes without intervening operations to flush or explicitly position the stream:</p>
<ul>
<li>after writing, call fseek(), fsetpos(), rewind(), or fflush() before reading</li>
<li>after reading, call fseek(), fsetpos(), or rewind() before writing unless the file is at EOF: a read that hits EOF can be followed immediately by a write</li>
</ul>
<p>The last thing I wanted to mention about reading and writing is a freak consequence of weird machines. On some digital signal processors (or more generally on the <a href="http://wikibin.org/articles/deathstation-9000.html">DeathStation 9000</a>), both char and int are 32 bits. This causes a vulnerability in the common pattern:</p>
<div>
<pre>
<code><span>int</span> c; <span>while</span> ((c = getchar()) != EOF) ...;</code>
</pre>
</div>
<p>There is no extra room in the int type to hold EOF as distinct from a valid character code, so a valid character can make the loop stop early. It&rsquo;s a high severity bug, a variation of which caused a nasty vulnerability in the bash shell, <a href="https://www-uxsup.csx.cam.ac.uk/pub/webmirrors/www.cert.org/advisories/CA-1996-22.html">CA-1996-22</a>.</p>
<p>Fine, you say, you don&rsquo;t plan to target such machines!</p>
<div>
<pre>
<code><span>#include </span><span>&lt;limits.h&gt;</span> <span>#if UCHAR_MAX == UINT_MAX</span> <span>#error "Your machine is weird."</span> <span>#endif</span></code>
</pre>
</div>
<p>Also you&rsquo;ll be careful not to indicate a parse failure with an unsigned code that casts to a signed value of EOF. Well the same logic can still trick you in C99 with wide characters if you&rsquo;re not careful.</p>
<p>It works like this: the fgetwc(), getwc(), and getwchar() functions return a value of type wint_t. This value can represent the next wide character read, or it can represent WEOF, which indicates end-of-file for wide character streams. On <em>most implementations</em>, the wchar_t type has the same width as wint_t, and these functions can return a character indistinguishable from WEOF. For this situation be sure to check after the loop for feof() and ferror(). If neither happened then you&rsquo;ve received a wide character that resembles WFEOF.</p>
<p>It&rsquo;s yet another place where wide character implementations are half baked.</p>
<h4 id="stdio.h-formatting">(stdio.h) formatting</h4>
<p>Scanf and printf have formatting options I hadn&rsquo;t seen before. First of all they can limit the length of strings they read or write.</p>
<div>
<pre>
<code><span>/* limit to 10 characters */</span> printf(<span>"%.10s"</span>, big_string); <span>/* or limit to n characters */</span> printf(<span>"%.*s"</span>, n, big_string); <span>/* limit to 10 characters */</span> scanf(<span>"%10s"</span>, input);</code>
</pre>
</div>
<p>I used to think scanf was very unsafe, but this limiting helps. Scanf also supports &ldquo;scansets&rdquo; to match strings containing specific characters. Here is how to match up to ten vowels: <code>%10[aeiou]</code>. Scanf also allows you to match but not capture, using *. E.g. <code>%*d</code>. The <code>%n</code> option saves the number of characters read so far in the scan.</p>
<p>Finally, in C99 printf has modifiers for size_t (%zu) and ptrdiff_t (%td). Because they are typedefs which change by architecture there is otherwise no way to specify them portably.</p>
<h3 id="stdlib.h">stdlib.h</h3>
<p>Stdlib is a hodgepodge. It has six categories of functions inside:</p>
<ul>
<li>algorithms (search, sort, rand)</li>
<li>integer functions</li>
<li>number parsing</li>
<li>multibyte conversions</li>
<li>storage allocation</li>
<li>environmental interactions</li>
</ul>
<h4 id="stdlib.h-random-numbers">(stdlib.h) random numbers</h4>
<p>Let&rsquo;s start with an interesting topic: random numbers. The standard library provides a rand() function to generate a pseudorandom sequence starting from a seed specified by srand(). The numbers range from 0 to RAND_MAX. The first problem is that RAND_MAX can be very small (~ 65535) on some platforms. Second problem is that the quality of rand() is not generally very good on most platforms. On Mac OS it is horrible. Changing the random seed by only a small amount (such as seeding by the epoch), leads to similar initial random values.</p>
<table>
<thead>
<tr>
<th>seed</th>
<th>1st rand</th>
</tr>
</thead>
<tbody>
<tr>
<td>1500000000</td>
<td>1189467867</td>
</tr>
<tr>
<td>1500000001</td>
<td>1189484674</td>
</tr>
<tr>
<td>1500000002</td>
<td>1189501481</td>
</tr>
<tr>
<td>1500000003</td>
<td>1189518288</td>
</tr>
<tr>
<td>1500000004</td>
<td>1189535095</td>
</tr>
<tr>
<td>1500000005</td>
<td>1189551902</td>
</tr>
<tr>
<td>1500000006</td>
<td>1189568709</td>
</tr>
<tr>
<td>1500000007</td>
<td>1189585516</td>
</tr>
<tr>
<td>1500000008</td>
<td>1189602323</td>
</tr>
<tr>
<td>1500000009</td>
<td>1189619130</td>
</tr>
</tbody>
</table>
<p>Rather than rely on whatever implementation of rand you get on a given architecture, it&rsquo;s just as easy to define your own xorshift rand function.</p>
<div>
<pre>
<code><span>static</span> <span>unsigned</span> <span>long</span> g_rand_state = <span>0</span>; <span>unsigned</span> <span>long</span> defensive_rand() { <span>uint32_t</span> x = g_rand_state; x ^= x &lt;&lt; <span>13</span>; x ^= x &gt;&gt; <span>17</span>; x ^= x &lt;&lt; <span>5</span>; <span>/* restrict to 32-bits */</span> x &amp;= <span>0xFFFFFFFF</span>; <span>return</span> (g_rand_state = x); }</code>
</pre>
</div>
<p>The constants 13, 17, 5 are specific to 32-bit numbers, so we mask the result to keep it within that range. There are also good 64-bit versions you can use with the &ldquo;long long&rdquo; type in C99. See also Chris Wellons&rsquo; <a href="https://nullprogram.com/blog/2018/07/31/">discussion</a> of creating good rand() functions through an exhaustive search.</p>
<p>Now we have a good rand() function, but how should we seed g_rand_state with an initial value? The typical way is to gather the epoch from <code>time(NULL)</code>, but it will mean the application will use the same seed if run more than once per second.</p>
<p>There are two other sources of entropy available from the standard library. One is hashing the path generated by tmpnam(), which in many implementations consults the process ID or a higher precision clock. The second is hashing the address of the main function, which is often fairly unpredictable due to address space layout randomization (ASLR). Do not rely on ASLR alone, however, as not all platforms use it. The address of main() should always be mixed with other sources of entropy.</p>
<p>Using the address of main numerically is a little tricky. C99 has a type called intptr_t which can hold pointer values as an integer, but this type is for data pointers only, not function pointers which on some architectures have a different size. We might consider casting a pointer to main as (char *) and reading the bytes, but for the same size reason this isn&rsquo;t feasible.</p>
<p>What we <em>can</em> do is create a function pointer to main, point another pointer at <em>it</em> and read the value out byte by byte. A pointer-to-function-pointer is just a data pointer and can be cast to (void *).</p>
<div>
<pre>
<code><span>int</span> (*p)(<span>int</span>, <span>char</span>**) = &amp;main; <span>unsigned</span> <span>char</span> bytes[<span>sizeof</span>(p) + <span>1</span>] = { <span>0</span> }; memcpy(bytes, (<span>void</span>*)&amp;p, <span>sizeof</span>(p));</code>
</pre>
</div>
<p>The bytes array is a NUL terminated string and can be hashed. To see these techniques in action and how to combine the entropy, check out <a href="https://github.com/begriffs/randln/blob/master/rand.c">rand.c</a>.</p>
<p>If you&rsquo;re willing to use functions beyond the C standard library, POSIX provides a random() function that is higher quality, and the BSDs/macOS/illumos (or Linux with libbsd) provide arc4random() which returns crypto-grade randomness.</p>
<h4 id="stdlib.h-integer-functions">(stdlib.h) integer functions</h4>
<p>In C89 the rounding of integer division isn&rsquo;t fully specified. When one of the numerator or denominator are negative it may either round toward zero or downward. It matches whatever the underlying hardware does. The committee didn&rsquo;t want to introduce overhead on any system going against the hardware convention. In C99 they changed their mind, reasoning that Fortran (known for numerical programming) defines the rounding.</p>
<p>C99 rounds toward zero. To match this behavior in C89, use the div and ldiv functions. They return structures div_t and ldiv_t that contain both the quotient and remainder of the division. The only reason to use these functions in C99 is for efficiency, because the functions may be implemented with a compiler builtin that can compute the quotient and remainder together in a single assembly instruction.</p>
<p>A note about the ato{i,l,f} functions &ndash; their behavior is <em>undefined</em> if the input cannot be parsed correctly. These functions need not set errno even. Except for behavior on error, atoi is equivalent to <code>(int)strtol(nptr, (char **)NULL, 10)</code>. The strto{d,l,ul} functions should always be preferred, because they provide proper error reporting and choice of base.</p>
<p>The implementation of strtod in Plauger&rsquo;s book was very careful about floating point overflow. It processed digits in groups of eight and then combined them later into a final result. It also consulted the decimal_point attribute set by the LC_MONETARY part of the current locale.</p>
<h4 id="stdlib.h.-memory">(stdlib.h.) memory</h4>
<p>An interesting bit of history: malloc used to be considered a low-level UNIX function, while calloc (&ldquo;C alloc&rdquo;) was conceived as the portable C version. However the committee standardized malloc too because it has less overhead (doesn&rsquo;t zero the memory). Nowadays it seems malloc is more popular. There also used to be a cfree, but it was identical to free and didn&rsquo;t make the cut for ANSI C.</p>
<p>The most flexible allocation function is realloc because it can simulate both malloc and free. When passed NULL for the original pointer it behaves like malloc, and when passed 0 for requested size it acts like free.</p>
<h4 id="stdlib.h-termination-and-execution">(stdlib.h) termination and execution</h4>
<p>Stdlib contains the EXIT_FAILURE and EXIT_SUCCESS macros with implementation defined integer values to indicate success or failure on exit. These values would be returned by main or passed to exit(). The C standard actually treats <code>return 0</code> in main or <code>exit(0)</code> specially, and maps it to whatever the system specific success code is. (Similar to how 0 in a pointer context is treated specially as the NULL value the way we talked about earlier.) Thus you don&rsquo;t need to include stdlib.h just to return successfully.</p>
<p>EXIT_FAILURE is necessary for portably indicating failure, though. The raw value 1 is considered successful on some platforms.</p>
<p>Speaking of exit(), it causes &ldquo;normal&rdquo; termination. It closes all open file handles, deletes files created by tmpfile(), and calls any handlers registered by atexit() in reverse order of their registration. The abort() function, on the other hand, causes immediate and &ldquo;abnormal&rdquo; termination. It needn&rsquo;t flush buffers, remove temp files, or close open streams. It can be canceled by catching SIGABRT. Aborting can be useful because it will cause a core dump if the OS is configured to save one. Thus the assert() function calls abort() to produce a core file to debug assertion failure.</p>
<p>Stdlib.h provides the system() function to run a command in the shell. If the command is a NULL pointer, system() will return non-zero if the command interpreter is available, and zero if it is not. The CERT standard says system() is a security violation and flat out says not to call it. It&rsquo;s easy to make a mistake with system(). Commands that are not fully pathed can be overridden, and even relative paths can be tricky if the attacker can control the current directory. And of course unsanitized input can attack through the shell. CERT suggests using execve() in POSIX to call fully pathed executables.</p>
<h4 id="stdlib.h-wchar_t">(stdlib.h) wchar_t</h4>
<p>OK, this is going to get complicated. The takeaway is that the C standard library cannot <em>portably</em> handle Unicode. You&rsquo;ll need a good third-party library. Let&rsquo;s see why this is.</p>
<p>Before locales and all that, C was designed for 7-bit ASCII text. The committee endorsed the use of locales to specify a codepage that reassigned the meaning of extra characters using all eight bits (the minimum allowed size of char). However some languages use more than 255 symbols (+ NUL). There were two ways to handle bigger alphabets: use bigger chunks of memory to hold each character code, or define special sequences of single byte characters to stand for extended characters.</p>
<p>These approaches are called &ldquo;wide characters&rdquo; and &ldquo;multibyte characters.&rdquo; Generally because networking equipment and disk storage is byte oriented, programs use multi-byte character encoding for communication with the outside world and storage. However for in-memory use people felt that using wide characters would be cleaner, because each cell in an array would map to exactly one character, like in the old ASCII string days.</p>
<p>The C89 standard is not very helpful with multibyte or wide character encoding. It merely set the stage with a wchar_t type for wide characters, and functions in stdlib.h to convert multibyte to wchar_t (mbstowcs) and the reverse (wcstombs) based on the locale. The committee was waiting to see how people wanted to work with international characters before standardizing it. K&amp;R 2nd edition does not list these functions but Plauger and the C89 spec have them.</p>
<p>The whole idea was that locale-specific multibyte to wide character conversion was supposed to be more general than any particular encoding system. However, nowadays Unicode has proven to be more popular than other encodings. The multibyte UTF-8 encoding is the standard interchange on the web, and the UCS-4 character set is up to the task of handling all the world&rsquo;s languages and then some. Should be no problem, right?</p>
<p>Well, at the time C99 was being standardized, the Unicode consortium (actually the contemporaneous European ISO committee) was endorsing the more limited Universal Coded Character Set UCS-2. The codepoints for UCS-2 are 16 bits, so that&rsquo;s the minimum width that C99 requires for wchar_t. Sadly the committee made their decision shortly before four byte UCS-4 was proposed (ISO 10646).</p>
<p>Vendors like Microsoft jumped into Unicode right away and implemented wchar_t as 16 bits. It&rsquo;s deep in their APIs and they&rsquo;re now stuck with that size for backward compatibility. Even Mac OS and iOS use 16 bit wchar_t for whatever reason.</p>
<p>UTF-16 combines the worst of multibyte characters and wide characters. a) Characters outside the &ldquo;base multilingual plane&rdquo; require two UTF-16 codepoints (called a surrogate pair) to represent them. This breaks the one-character-one-codepoint assumption. b) It&rsquo;s wasteful of memory because even ASCII characters take two bytes.</p>
<p>C99 provides wide character versions of the ctype functions, in &lt;wctype.h&gt;, but they simply cannot work properly with surrogate pairs. For example (as pointed out <a href="https://lists.gnu.org/archive/html/bug-gnulib/2011-02/msg00013.html">here</a>):</p>
<ul>
<li>0xD800 0xDF1E = U+1031E is a letter (iswalpha should be true)</li>
<li>0xD800 0xDF20 = U+10320 is not a letter (iswalpha should be false)</li>
<li>0xD834 0xDF1E = U+1D31E is not a letter (iswalpha should be false)</li>
<li>0xD834 0xDF20 = U+1D320 is not a letter (iswalpha should be false)</li>
<li>0xD835 0xDF1E = U+1D71E is a letter (iswalpha should be true)</li>
<li>0xD835 0xDF20 = U+1D720 is a letter (iswalpha should be true)</li>
</ul>
<p>Neither the first nor the second element of the pair alone can predict whether the resulting Unicode character is alphabetic. There is no way that a system can provide this information through a function &lsquo;iswalpha&rsquo; that takes a single wchar_t argument.</p>
<p>C99 does guarantee a macro will be present when the current environment is ISO 10646 compliant, meaning wchar_t can hold every UCS-4 codepoint. We can blow up for the other platforms:</p>
<div>
<pre>
<code><span>#ifndef __STDC_ISO_10646__</span> <span>#error "Your wide characters suck."</span> <span>#endif</span></code>
</pre>
</div>
<p>Even assuming we restrict the code to ISO 10646 systems only, the wctype functions are too crude to deal with the subtleties of international languages. Because of how Unicode characters can join together, it&rsquo;s infeasible to use pointer arithmetic to calculate &ldquo;string length&rdquo; or parse &ldquo;words&rdquo; with iswspace() robustly.</p>
<p>Some parts of programs can continue to operate on text as an opaque series of bytes. However for other parts that must inspect the characters themselves, you should use a sophisticated Unicode library like <a href="http://site.icu-project.org/">ICU</a> or <a href="http://juliastrings.github.io/utf8proc/">utf8proc</a>. This has the advantage of working with C89, so you won&rsquo;t be forced to upgrade to C99 just because of text processing.</p>
<p>With a good Unicode library we don&rsquo;t need wide characters. We can use UTF-8 everywhere, even in program memory. That&rsquo;s the school of thought behind <a href="https://utf8everywhere.org">utf8everywhere.org</a>.</p>
<h4 id="stdlib.h-lessons-from-the-code">(stdlib.h) lessons from the code</h4>
<p>While reading Plauger&rsquo;s implementation for stdlib, I noticed some tricks worth sharing.</p>
<p>The comma operator can be used to group multiple small assignments together under an if statement without needing to add curly braces.</p>
<div>
<pre>
<code><span>if</span> (condition) foo = <span>1</span>, bar = <span>2</span>;</code>
</pre>
</div>
<p>Also an if statement with no statements can be used to shrink a big expression in another if statement:</p>
<div>
<pre>
<code><span>if</span> (cond) ; <span>else</span> <span>if</span> (big cond) foo;</code>
</pre>
</div>
<p>C89 allows bare blocks inside a function to segregate variable declarations near the code that uses them. The variables are not accessible outside the block. It can help readability to know when variables are no longer needed, although you might argue that it suggests the function as a whole is too big.</p>
<div>
<pre>
<code><span>#include </span><span>&lt;stdio.h&gt;</span> <span>int</span> main(<span>void</span>) { puts(<span>"Hello."</span>); { <span>int</span> i = <span>0</span>; printf(<span>"%d</span><span>\n</span><span>"</span>, i); } <span>/* error: use of undeclared identifier 'i' */</span> printf(<span>"%d</span><span>\n</span><span>"</span>, i); }</code>
</pre>
</div>
<p>The trick is even useful in C99 to reuse and reinitialize variables across repetitive test cases in a single function.</p>
<p>Speaking of brackets, Plauger uses Whitesmiths style indentation, with brackets indented to the level of their code:</p>
<div>
<pre>
<code><span>if</span> (cond) { foo(); <span>while</span> (cond) { bar(); baz(); } }</code>
</pre>
</div>
<p>I find it difficult to read (probably just unfamiliar), but it does have an internal consistency. The brackets are wrapping up multiple statements into a single unit, and this unit is indented the same way that a single statement would be. Still not going to indent this way, but just saying.</p>
<p>Another interesting trick is negating an unsigned number. Plauger does that to consolidate code for signed and unsigned numbers in the same function.</p>
<pre>
<code>unsigned long _Stoul(const char *, char **, int); #define atoi(s) (int)_Stoul(s, NULL, 10); #define atol(s) (long)_Stoul(s, NULL, 10); #define strtoul(s, endptr, base) _Stoul(s, endptr, base);</code>
</pre>
<p>This _Stoul() function negates its unsigned long value if the string it is parsing has a negative sign. This operates bitwise on an unsigned value the same as it would on signed, and after casting for atoi and atol it will be negative as expected. I didn&rsquo;t know it was &ldquo;allowed,&rdquo; but C doesn&rsquo;t care.</p>
<h3 id="string.h">string.h</h3>
<p>This header is divided between functions starting with str- and those starting with mem-. The former work with NUL terminated strings, and the latter operate with explicit lengths. Some of the str- functions have a modifier to limit length (strncat, strncmp, strncpy) or a modifier to work backward (strrchr). However the header doesn&rsquo;t have all permutations. For instance no strnlen or memrchr.</p>
<h4 id="string.h-widened-types">(string.h) widened types</h4>
<p>Why does strchr take an int rather than a char for its search character? Same with memset, it takes an int value, but converts it to unsigned char internally. The spec dictates this in section 7.21.6.1. That misleading int signature is for backward compatibility with pre-ANSI code which lacked function prototypes and promoted char arguments to int. Under default argument promotions any integral type smaller than int (or unsigned int) always converts to int.</p>
<p>All standard library functions are specified in terms of the &ldquo;widened&rdquo; types. It&rsquo;s just that string.h contains many functions where it is especially apparent. The widened types ensure that most library functions can be called with or without a prototype in scope. Legacy code doesn&rsquo;t use prototypes, and ANSI C did not want to break backward compatibility. Even if people were willing to update all the legacy code, any legacy modules distributed as compiled object files rather than source would not link properly against functions with changed argument types.</p>
<p>Similar rationale lies behind the standard&rsquo;s guarantee that char pointers and void pointers share the same representation and alignment requirements. Relying on this guarantee allows old code that to work with the void* malloc in place of the original char* malloc. (See the C99 Rationale section 7.1.4, Use of Library Functions.)</p>
<h4 id="string.h-signed-vs-unsigned">(string.h) signed vs unsigned</h4>
<p>We&rsquo;re not done with the type mysteries in string.h. Why is it that strchr casts its int argument to char internally and memchr casts to unsigned int? Well given that strchr is searching through char*, it makes sense to match the type. But memchr is searching through void*. Why not compare each memory location with char rather than unsigned char?</p>
<p>It turns out that the C standard makes special guarantees about unsigned char that make it an ideal type to represent arbitrary binary information.</p>
<ol>
<li>Unsigned char is guaranteed to have no padding bits. All bits contribute to the value of the data. Other types on some architectures include things like parity bits which don&rsquo;t affect the value itself but do use space.</li>
<li>No bitwise operation starting from an unsigned char value, when converted back into that type, can produce overflow, trap representations or undefined behavior. It can be freely manipulated. Trap representations are certain bit patterns that are reserved for exceptional circumstances, like the NaN value in floating point numbers.</li>
<li>Accessing parts of a larger data object ith an unsigned char pointer will not violate any &ldquo;aliasing rules.&rdquo; The unsigned char pointer will be guaranteed to see all modifications of the data object.</li>
</ol>
<p>In &ldquo;Portable C,&rdquo; Rabinowitz says the char type has a few distinct uses: codeset characters, units of storage, small numbers, and small bit patterns. He recommends regular char for codepoints and small numbers, and unsigned for the others. (In C99 wchar_t might be considered the proper way to represent codeset characters, but we&rsquo;ve already seen the difficulty there.)</p>
<p>Note that Rabinowitz doesn&rsquo;t specify signed char and unsigned char, but rather plain char and unsigned char. That&rsquo;s because C does not specify whether plain char is signed or unsigned. We would get a warning trying to pass a definitively signed char* to a function in string.h if the default was unsigned char for that platform.</p>
<h4 id="string.h-mem--implementations">(string.h) mem- implementations</h4>
<p>The next functions that taught me something are memcpy and memmove. The first one blasts bytes from one part of memory to another, possibly taking advantage of machine instructions to do large block copies. It doesn&rsquo;t check for an overlap between the source and destination, in which situation the results are undefined. C99 marks the source and destination pointers with the <code>restrict</code> qualifier to allow the compiler to optimize under that assumption.</p>
<p>Memmove is the slower more careful brother. It works correctly even if the source and destination areas overlap. It is specified to act <em>as if</em> the source memory is first copied to a separate buffer, then copied into the destination. When I tried writing it that is exactly what I did, but Plauger has a faster way:</p>
<div>
<pre>
<code><span>void</span> *(memmove)(<span>void</span> *d, <span>const</span> <span>void</span> *s, <span>size_t</span> n) { <span>unsigned</span> <span>char</span> *ds = d; <span>const</span> <span>unsigned</span> <span>char</span> *ss = s; <span>if</span> (s &gt; d) <span>while</span> (n-- &gt; <span>0</span>) *ds++ = *ss++; <span>else</span> <span>for</span> (ss += n, ds += n; <span>0</span> &lt; n; --n) *ds-- = *ss--; <span>return</span> d; }</code>
</pre>
</div>
<p>This compares the pointer positions to see which occurs before the other in memory, then does the copy from left to right or right to left depending on which comes before the other. This is very fast and uses no extra space.</p>
<p>However, isn&rsquo;t comparing random pointers undefined behavior? C allows you to compare pointers within the same primary data object (like the addresses of different cells in the same array), but not any random pointers. Objects could be in different segments of a segmented memory architecture, or even in totally different memory banks such as in a Harvard architecture.</p>
<p>But with memmove why would you call it when copying one data object into a totally different one? There would be no danger that they overlap, unless you are copying too many bytes, which is already its own big problem. Thus we can assume that the pointers to memmove <em>are</em> in the same data object and thus comparable. What this also tells me is not to indiscriminately use memmove all the time in order to be &ldquo;safe&rdquo; or something, because some implementations like Plauger&rsquo;s would then cause undefined behavior.</p>
<p>One other thing to note about the memmove implementation is that it uses unsigned char pointers to do the work rather than void pointers. Doing pointer arithmetic on void* is a GNU-ism not permitted in portable C.</p>
<h4 id="string.h-strncpy">(string.h) strncpy</h4>
<p>No discussion of string.h is complete without talking about the dangers of that &ldquo;hybrid&rdquo; function strncpy(). I call it a hybrid because in some ways it acts like a mem- function operating on blocks of memory. Let&rsquo;s look at the history.</p>
<p>The function was originally used to copy filenames into fixed width fields. Two fields were considered equal when each and every character of one matched the other. The comparison didn&rsquo;t stop at <code>\0</code> but blindly compared the full length of both memory regions. To represent a short file name, the remainder of the memory was padded with <code>\0</code> just to normalize it. A long filename that used every available character wouldn&rsquo;t have a <code>\0</code> at all because the length was already known.</p>
<p>So while strncpy might look like a &ldquo;safe&rdquo; strcpy that is guaranteed not to overflow a destination buffer, it&rsquo;s actually pretty weird and not safe. Consider this snippet:</p>
<div>
<pre>
<code><span>char</span> dst[BUFSZ], *src = <span>/* ... */</span>; strncpy(dst, src, BUFSZ);</code>
</pre>
</div>
<p>It copies the first BUFSZ characters of src into dst. If BUFSZ is shorter than src (meaning <code>\0</code> is not one of the first BUFSZ characters) then <code>\0</code> doesn&rsquo;t get written to dst! The result is not NUL terminated. Also, strncpy does a bunch of potentially useless work adding extra NUL padding if BUFSZ is larger than strlen(src).</p>
<p>The proper way to handle strncpy depends on how you want to treat string truncation. In some cases it would be an error to truncate the source string, such as if it holds a URL or filename. In this case the code should check for truncation.</p>
<div>
<pre>
<code><span>if</span> (strlen(src) &lt; BUFSZ) strcpy(dst, src); <span>/* a regular strcpy is safe */</span> <span>else</span> <span>/* handle truncation error */</span></code>
</pre>
</div>
<p>If truncation <em>is</em> permissible, then be sure to include a NUL character in the destination.</p>
<div>
<pre>
<code><span>/* either of these techniques work in C89: */</span> <span>/* adding the NUL character manually */</span> strncpy(dst, src, BUFSZ); dst[BUFSZ-<span>1</span>] = '\<span>0</span>'; <span>/* or truncating dst and using (safe) strncat */</span> dst[<span>0</span>] = '\<span>0</span>'; strncat(dst, src, BUFSZ); <span>/* C99 provides snprintf that can do this in one line */</span> snprintf(dst, BUFSZ, <span>"%s"</span>, src);</code>
</pre>
</div>
<p>The C11 standard in Annex K contains functions with &ldquo;bounds-checking&rdquo; interfaces. This includes strncpy_s which acts more safely. The functions in Annex K are safe for other reasons as well, but covering this goes too far afield from the C89 focus of this article. One thing to note is that it provides a memset_s which is guaranteed not to be removed by an optimizing compiler. Sometimes memset is optimized out if it operates on memory that the compiler can tell is not read afterwards.</p>
<p>Those are the notes I wanted to share for this header. Also C99 introduces wide character versions of these functions in wchar.h.</p>
<h3 id="time.h">time.h</h3>
<p>First, the C standard is very lenient about this header. It has functions to do all kinds of conversions, but the bigger picture is that implementations are allowed to make their &ldquo;best approximation&rdquo; to the date and time. Some of them might do a bad job and yet still conform to the standard. Many C environments do not support the concepts of daylight savings or time zones. Both notions are defined geographically and politically, and thus may require more knowledge about the real world than an implementation can support.</p>
<p>There are a lot of details in this header that I don&rsquo;t want to regurgitate, but it is useful to see how the functions convert between the data types. I made a graph to make it clearer.</p>
<div>
<img src="https://begriffs.com/images/time-h.gif" alt="functions in time.h">
<p>functions in time.h</p>
</div>
<p>One thing I didn&rsquo;t know was that the header provides a clock() function to measure the CPU time elapsed since program start. Also that it provides a resolution of CLOCKS_PER_SEC which is often higher than one. The rest of the library is limited to nothing smaller than seconds of precision.</p>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via begriffs.com https://begriffs.com<br>
<br>
June 11, 2019 at 02:33PM
